Fix compilation on Ubuntu 16.04.

mychele opened the pull request:
With the default ns-3 waf settings the compilation is not successful in Ubuntu 16.04, due to two warnings which are treated as errors:
1) the order in which variables are initialized in one of the two mptcp-socket-base.cc constructors
2) unused variables in tcp-socket-base.cc

Moreover, it would be nice to specify somewhere (for example, in the wiki) that, for Ubuntu, libgcrypt20-dev and libssl-dev must be installed, and that the CXXFLAG -std=c++14 should be used on configuration.

mychele committed:
Fix compilation on Ubuntu 16.04. libgcrypt20-dev and libssl-dev must be installed, moreover configure with the CXXFLAG -std=c++14

teto commented:
Would you mind removing the comments `// michele@polese.io` ? Your name will appear in the git history so you will get credit anyway :)
I don't use any C++14 feature, only C++11, I will mention it on the wiki.
Does the build fail if libgcrypt is not present ? I sent a patch upstream but it may not be the case in this branch. Anyway it's best if available so I will add it to the wiki again.
I use ubuntu 16.04 as well, I may have disabled warning as error (which is a pain when developing sthg), I will check this evening.
Thanks for the patch

mychele commented:
- I will remove it, as soon as I can I will commit the code without comment! :)
- I think that C++11 and C++14 are equivalent for your code, since std::aligned_storage is introduced in C++11, but by default ns-3 is not compiled with neither of the two.
- As for libgcrypt20-dev, the build does not fail if it is not installed. Instead, it fails if libssl-dev is not installed (it misses openssl/sha.h header). However, there is something weird going on. I have libgcrypt20 already installed. If I install libgcrypt20-dev, when calling ./waf configure the gcrypt library is enabled, otherwise it is not. In both cases, the scratch and examples folders are not built, i.e., in the build folder scratch and examples folders are not present! If I comment line 697 in the wscript file (program.use.append('GCRYPT')) then they are compiled (well, actually the compilation of the examples fails because of errors in mptcp.cc, but at least examples are considered for compilation). Any idea of what is going on? I commented the line and built the tests, the ones related to mptcp pass.

teto commented:
I used openssl in a first implementation then switched to libgcrypt due to license concerns.
Inclusion of openssl/sha.h should be removed.
The gcrypt linking was made in a hurry but the patch I sent upstream is much cleaner so I am fine leaving as is.
If there is no scratch folder, then I should add it. You have to configure with --enable-examples to enable examples

mychele commented:
What I meant is that despite the presence of scratch folder and the --enable-examples flag, these were not compiled. I replaced the wscript with the one in your master branch and update waf to 1.8.19 as in your master branch, and this problem disappears. 
I also removed the openssl references, if you want I can commit also this modification.

teto commented:
Nice please do

mychele committed:
Update wscript and waf with master branch ones. Fix compilation errors

teto commented:
Could you post the waf backport in a different PR ?  the wscript contains trailing spaces which should not be the case, are you sure your editor doesn't do funky stuff with files ?

Also git blame says I haven't touched src/internet/test/tcp-header-test.cc yet you modify it. upstream should compile correctly so that's weird. what's your compiler (version) and how do you configure ns3 ?

mychele committed:
Revert waf and wscript to the ones of merge_nat branch

mychele commented:
I am using Ubuntu 16.04 with gcc 5.4.0 20160609.
Actually, it seems you made some additions to tcp-header-test.cc: 
https://github.com/teto/ns-3-dev-git/commits/merge_nat/src/internet/test/tcp-header-test.cc
As I was mentioning, I updated the wscript because the scratch folder and the examples were not compiled, even if the scratch folder is present and I used the --enable-examples flag
